pkg/proxy/proxy.go

```diff
@@ -264,47 +267,208 @@ func (p *Proxy) health(ctx echo.Context) error {
+	if context.IsPublicViewerEnabled(ctx) {
+		return nil
+	}
+
+	// retrieve the UserSignup for the requesting user.
+	//
+	// UserSignup complete status is not checked, since it might cause the proxy blocking the request
+	// and returning an error when quick transitions from ready to provisioning are happening.
+	userSignup, err := p.app.SignupService().GetSignupFromInformer(nil, userID, username, false)
+	if err != nil {
+		return err
+	}
+
+	// if the UserSignup is nil or has NOT the CompliantUsername set,
+	// it means that MUR was NOT created and useraccount is NOT provisioned yet
+	// TODO(@filariow): should we also check if the user is Banned?
+	if userSignup == nil || userSignup.CompliantUsername == "" {
+		cause := errs.New("user is not provisioned (yet)")
+		log.Error(nil, cause, fmt.Sprintf("signup object: %+v", userSignup))
+		return cause
+	}
+	return nil
+}
+
+// getClusterAccess retrieves the access to the cluster hosting the requested workspace,
+// if the user has access to it.
+// Access can be either direct (a SpaceBinding linking the user to the workspace exists)
+// or community (a SpaceBinding linking the PublicViewer user to the workspace exists).
+func (p *Proxy) getClusterAccess(ctx echo.Context, userID, username, workspaceName, proxyPluginName string, workspace *toolchainv1alpha1.Workspace) (*access.ClusterAccess, error) {
+	// retrieve by name the workspace from the list of workspaces the user has access to.
+	if workspace == nil || workspace.Name != workspaceName {
+		return nil, crterrors.NewInternalError(errs.New("unable to get target cluster"), "workspace not found")
+	}
+
+	// retrieve cluster access as requesting user or PublicViewer
+	cluster, err := p.getClusterAccessAsUserOrPublicViewer(ctx, userID, username, proxyPluginName, workspace)
+	if err != nil {
+		return nil, crterrors.NewInternalError(errs.New("unable to get target cluster"), err.Error())
+	}
+	return cluster, nil
+}
+
+// getClusterAccessAsUserOrPublicViewer if the requesting user has direct access to the workspace,
+// it returns the ClusterAccess impersonating the requesting user.
+// Otherwise, if PublicViewer support is enabled and PublicViewer user has access to the workspace,
+// it returns the ClusterAccess impersonating the PublicViewer user.
+func (p *Proxy) getClusterAccessAsUserOrPublicViewer(ctx echo.Context, userID, username, proxyPluginName string, workspace *toolchainv1alpha1.Workspace) (*access.ClusterAccess, error) {
+	// retrieve the requesting user's UserSignup
+	userSignup, err := p.spaceLister.GetSignupFunc(nil, userID, username, false)
+	if err != nil {
+		return nil, crterrors.NewInternalError(errs.New("unable to get target cluster"), "user not found")
+	}
+
+	// proceed as PublicViewer if needed
+	publicViewerEnabled := context.IsPublicViewerEnabled(ctx)
+	if publicViewerEnabled && !hasDirectAccess(userSignup, workspace) {
```

Comment:
Let's maybe add a comment stating that we attempt to access the workspace as a public viewer only if the user doesn't have any role in the workspace. So we always access the workspace using the most powerful role available to the user: the one which is specified by the space binding. And only if there is none then we use the public viewer user. We also do not check if the public viewer actually has any access in the workspace. If there is no role for the public viewer then the request will be rejected by the member cluster.

Reply:
done in 048f29ba1bc98e564f62713f6bc277a8d2e90e09

Reply:
Also, I think we should check if public-viewer has access before forwarding the request. So this is a gap I need to fill.

Reply:
done in 6408242518acad1da5c980985948e8af5dd8b25f

Reply:
> Also, I think we should check if public-viewer has access before forwarding the request. So this is a gap I need to fill.

It would work without that check too. We would forward the request but the request would be rejected by the api server. But this check will reject the request faster (so it will consume less resources). So, yeah, we can have this check here.